operator/apis/resources/v5/role_types.go

```diff
@@ -28,7 +28,7 @@ func init() {
 }
 
 // TeleportRoleSpec defines the desired state of TeleportRole
-type TeleportRoleSpec types.RoleSpecV5
+type TeleportRoleSpec types.RoleSpecV6
```

Comment:
Looking at the file path, it's under `apis/resources/v5` directory. Should this be moved to `resources/v6` instead?

Reply:
The `Role` resource also exists in `resources/v6`, and we need to continue to serve the resource through the v5 API for at least a major version. A v5 roleSpec deserializes nicely in a v6 roleSpec, so I don't see a problem with using the v6 roleSpec instead as long as the case where `kubernetes_resource: []` is handled in a backward-compatible fashion for v5.

However, I'm curious how multiple versions of the same resource will be handled. I suspect the behaviour will not be adequate unless we do a custom conversion. Let's say:
- I run teleport-operator v11
- I create a role using the v5 API (serialized in etcd as v5)
- I upgrade to v12
- I get the role through the v5 API:
   ```
   apiVersion: resources.teleport.dev/v5
   kind: TeleportRole
   metadata:
     name: myrole
   spec:
     allow:
       kubernetes_labels:
         "*": "*"
       kubernetes_groups: ["kube_group"]
   ```
- I get the same role through the v6 API (Kube will apply the default `None` conversion strategy):
   ```
   apiVersion: resources.teleport.dev/v6
   kind: TeleportRole
   metadata:
     name: myrole
   spec:
     allow:
       kubernetes_labels:
         "*": "*"
       kubernetes_groups: ["kube_group"]
   ```

The two specs are the same, but the behaviour is different. The v6 role equivalent to the v5 I previously created should be:
```
apiVersion: resources.teleport.dev/v6
kind: TeleportRole
metadata:
  name: myrole
spec:
  allow:
    kubernetes_labels:
      "*": "*"
    kubernetes_groups: ["kube_group"]
    kubernetes_resources:
    - kind: pod
      name: *
      namespace: *
   ```

So depending on the API used to request the resource, I receive two descriptions of the same resource with different behaviours. I believe the same problem will happen with resource creation and edition. I will run more tests, but I think `conversion: None` will not be sufficient in our case.

I will run more tests next week to understand precisely how versions are managed at the operator level.

Reply:
I think we need to implement the [`Hub`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/conversion#Hub) interface for Role V5 and the [`Convertible`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/conversion?tab=doc#Convertible) (Spoke) interface for v5.

Resource controller converts from v5 -> v6 -> v5.
`v5`->`v6`: converted in the resource handler and the role controller only receives rolev6 objects.
`v6`->`v5`: converted after the controller handler finishes and before updating the resource in K8S.

I guess this is what happens but not sure